File: FBControlCore/Utility/FBVideoFileWriter.m
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

#import "FBVideoFileWriter.h"

#import <AVFoundation/AVFoundation.h>

#import <FBControlCore/FBControlCore.h>

#import "FBControlCoreError.h"

@interface FBVideoFileWriter () <AVCaptureFileOutputRecordingDelegate>

@property (nonatomic, readonly, strong) AVCaptureSession *session;
@property (nonatomic, readonly, strong) AVCaptureMovieFileOutput *output;
@property (nonatomic, readonly, strong) id<FBControlCoreLogger> logger;
@property (nonatomic, readonly, strong) FBMutableFuture<NSNull *> *startFuture;
@property (nonatomic, readonly, strong) FBMutableFuture<NSNull *> *finishFuture;
@property (nonatomic, readonly, copy) NSString *filePath;

@end

@implementation FBVideoFileWriter

+ (nullable instancetype)writerWithSession:(AVCaptureSession *)session filePath:(NSString *)filePath logger:(id<FBControlCoreLogger>)logger error:(NSError **)error
{
  // Add the Output to the Session.
  AVCaptureMovieFileOutput *output = [[AVCaptureMovieFileOutput alloc] init];
  if (![session canAddOutput:output]) {
    return [[FBControlCoreError
             describe:[NSString stringWithFormat:@"Cannot add File Output to session for %@", filePath]]
            fail:error];
  }
  [session addOutput:output];

  return [[FBVideoFileWriter alloc] initWithSession:session output:output filePath:filePath logger:logger];
}

- (instancetype)initWithSession:(AVCaptureSession *)session output:(AVCaptureMovieFileOutput *)output filePath:(NSString *)filePath logger:(id<FBControlCoreLogger>)logger
{
  self = [super init];
  if (!self) {
    return nil;
  }

  _session = session;
  _output = output;
  _filePath = filePath;
  _logger = logger;
  _startFuture = FBMutableFuture.future;
  _finishFuture = FBMutableFuture.future;

  return self;
}

#pragma mark Public Methods

- (FBFuture<NSNull *> *)startRecording
{
  NSError *innerError = nil;
  if ([NSFileManager.defaultManager fileExistsAtPath:self.filePath]) {
    [self.logger log:[NSString stringWithFormat:@"File already exists at %@, deleting", self.filePath]];
    if (![NSFileManager.defaultManager removeItemAtPath:self.filePath error:&innerError]) {
      return [[[FBControlCoreError
                describe:[NSString stringWithFormat:@"Failed to remove existing device video at %@", self.filePath]]
               causedBy:innerError]
              failFuture];
    }
    [self.logger log:[NSString stringWithFormat:@"Removed video file at %@", self.filePath]];
  }
  if (![NSFileManager.defaultManager createDirectoryAtPath:self.filePath.stringByDeletingLastPathComponent withIntermediateDirectories:YES attributes:nil error:&innerError]) {
    return [[[FBControlCoreError
              describe:[NSString stringWithFormat:@"Failed to remove create auxillary directory for device at %@", self.filePath]]
             causedBy:innerError]
            failFuture];
  }
  NSURL *file = [NSURL fileURLWithPath:self.filePath];
  [self.session startRunning];
  [self.output startRecordingToOutputFileURL:file recordingDelegate:self];
  [self.logger log:[NSString stringWithFormat:@"Started Video Session for Device Video at file %@", self.filePath]];
  return self.startFuture;
}

- (FBFuture<NSNull *> *)stopRecording
{
  [self.output stopRecording];
  [self.session stopRunning];
  return self.finishFuture;
}

- (FBFuture<NSNull *> *)completed
{
  return self.finishFuture;
}

#pragma mark Recording Delegate

- (void)captureOutput:(AVCaptureFileOutput *)captureOutput didStartRecordingToOutputFileAtURL:(NSURL *)fileURL fromConnections:(NSArray *)connections
{
  [self.logger log:[NSString stringWithFormat:@"Did Start Recording at %@", self.filePath]];
  [self.startFuture resolveWithResult:NSNull.null];
}

- (void)captureOutput:(AVCaptureFileOutput *)captureOutput didPauseRecordingToOutputFileAtURL:(NSURL *)fileURL fromConnections:(NSArray *)connections
{
  [self.logger log:[NSString stringWithFormat:@"Did Pause Recording at %@", self.filePath]];
}

- (void)captureOutput:(AVCaptureFileOutput *)captureOutput didFinishRecordingToOutputFileAtURL:(NSURL *)outputFileURL fromConnections:(NSArray *)connections error:(NSError *)error
{
  [self.logger log:[NSString stringWithFormat:@"Did Finish Recording at %@", self.filePath]];
  [self.finishFuture resolveWithResult:NSNull.null];
}

- (void)captureOutput:(AVCaptureFileOutput *)captureOutput didResumeRecordingToOutputFileAtURL:(NSURL *)fileURL fromConnections:(NSArray *)connections
{
  [self.logger log:[NSString stringWithFormat:@"Did Resume Recording at %@", self.filePath]];
}

- (void)captureOutput:(AVCaptureFileOutput *)captureOutput willFinishRecordingToOutputFileAtURL:(NSURL *)fileURL fromConnections:(NSArray *)connections error:(NSError *)error
{
  [self.logger log:[NSString stringWithFormat:@"Will Finish Recording at %@", self.filePath]];
}

@end

File: FBControlCore/Utility/FBVideoFileWriter.swift
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import AVFoundation
import Foundation

@objc(FBVideoFileWriter)
public class FBVideoFileWriter: NSObject, AVCaptureFileOutputRecordingDelegate {

  // MARK: Properties

  private let session: AVCaptureSession
  private let output: AVCaptureMovieFileOutput
  private let logger: any FBControlCoreLogger
  private let outputURL: URL
  private let lifecycleLock = NSLock()
  private var hasStarted = false
  private var hasFinished = false
  private var startAwaiters: [CheckedContinuation<Void, Error>] = []
  private var finishAwaiters: [CheckedContinuation<Void, Error>] = []

  // MARK: Initializers

  @objc(writerWithSession:filePath:logger:error:)
  public class func writer(withSession session: AVCaptureSession, filePath: String, logger: any FBControlCoreLogger) throws -> Self {
    let output = AVCaptureMovieFileOutput()
    if !session.canAddOutput(output) {
      throw FBControlCoreError.describe("Cannot add File Output to session for \(filePath)").build()
    }
    session.addOutput(output)
    return self.init(session: session, output: output, filePath: filePath, logger: logger)
  }

  required init(session: AVCaptureSession, output: AVCaptureMovieFileOutput, filePath: String, logger: any FBControlCoreLogger) {
    self.session = session
    self.output = output
    self.outputURL = URL(fileURLWithPath: filePath)
    self.logger = logger
    super.init()
  }

  // MARK: Public Methods

  public func start() async throws {
    try await startWriting()
  }

  public func stop() async throws -> URL {
    try await stopWriting()
    return outputURL
  }

  // MARK: Private Methods

  private var filePath: String {
    outputURL.path
  }

  private func startWriting() async throws {
    if FileManager.default.fileExists(atPath: filePath) {
      logger.log("File already exists at \(filePath), deleting")
      do {
        try FileManager.default.removeItem(atPath: filePath)
      } catch {
        throw FBControlCoreError.describe("Failed to remove existing device video at \(filePath)").caused(by: error).build()
      }
      logger.log("Removed video file at \(filePath)")
    }
    do {
      try FileManager.default.createDirectory(atPath: (filePath as NSString).deletingLastPathComponent, withIntermediateDirectories: true, attributes: nil)
    } catch {
      throw FBControlCoreError.describe("Failed to remove create auxillary directory for device at \(filePath)").caused(by: error).build()
    }
    try await withCheckedThrowingContinuation { (continuation: CheckedContinuation<Void, Error>) in
      registerStartAwaiter(continuation)
      session.startRunning()
      output.startRecording(to: outputURL, recordingDelegate: self)
      logger.log("Started Video Session for Device Video at file \(filePath)")
    }
  }

  private func stopWriting() async throws {
    try await withCheckedThrowingContinuation { (continuation: CheckedContinuation<Void, Error>) in
      registerFinishAwaiter(continuation)
      output.stopRecording()
      session.stopRunning()
    }
  }

  private func registerStartAwaiter(_ continuation: CheckedContinuation<Void, Error>) {
    lifecycleLock.lock()
    if hasStarted {
      lifecycleLock.unlock()
      continuation.resume()
    } else {
      startAwaiters.append(continuation)
      lifecycleLock.unlock()
    }
  }

  private func registerFinishAwaiter(_ continuation: CheckedContinuation<Void, Error>) {
    lifecycleLock.lock()
    if hasFinished {
      lifecycleLock.unlock()
      continuation.resume()
    } else {
      finishAwaiters.append(continuation)
      lifecycleLock.unlock()
    }
  }

  private func markStarted() -> [CheckedContinuation<Void, Error>]? {
    lifecycleLock.lock()
    defer { lifecycleLock.unlock() }
    if hasStarted {
      return nil
    }
    hasStarted = true
    let awaiters = startAwaiters
    startAwaiters = []
    return awaiters
  }

  private func markFinished() -> [CheckedContinuation<Void, Error>]? {
    lifecycleLock.lock()
    defer { lifecycleLock.unlock() }
    if hasFinished {
      return nil
    }
    hasFinished = true
    let awaiters = finishAwaiters
    finishAwaiters = []
    return awaiters
  }

  // MARK: AVCaptureFileOutputRecordingDelegate

  public func fileOutput(_ output: AVCaptureFileOutput, didStartRecordingTo fileURL: URL, from connections: [AVCaptureConnection]) {
    logger.log("Did Start Recording at \(filePath)")
    markStarted()?.forEach { $0.resume() }
  }

  public func fileOutput(_ output: AVCaptureFileOutput, didPauseRecordingTo fileURL: URL, from connections: [AVCaptureConnection]) {
    logger.log("Did Pause Recording at \(filePath)")
  }

  public func fileOutput(_ output: AVCaptureFileOutput, willFinishRecordingTo outputFileURL: URL, from connections: [AVCaptureConnection], error: Error?) {
    logger.log("Will Finish Recording at \(filePath)")
  }

  public func fileOutput(_ output: AVCaptureFileOutput, didFinishRecordingTo outputFileURL: URL, from connections: [AVCaptureConnection], error: Error?) {
    logger.log("Did Finish Recording at \(filePath)")
    markFinished()?.forEach { $0.resume() }
  }

  public func fileOutput(_ output: AVCaptureFileOutput, didResumeRecordingTo fileURL: URL, from connections: [AVCaptureConnection]) {
    logger.log("Did Resume Recording at \(filePath)")
  }
}
